File: src/aoc2015/Day04.java
package aoc2015;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

class Day04 {

  private static long partOne(String s) throws NoSuchAlgorithmException {
    return solve(s, "00000");
  }

  private static long partTwo(String s) throws NoSuchAlgorithmException {
    return solve(s, "000000");
  }

  private static long solve(String s, String difficulty) throws NoSuchAlgorithmException {
    MessageDigest digest = MessageDigest.getInstance("md5");
    String result;
    long cnt = 0;
    do {
      byte[] inputBytes = (s + cnt++).getBytes();
      byte[] hashBytes = digest.digest(inputBytes);
      result = javax.xml.bind.DatatypeConverter.printHexBinary(hashBytes);
    } while (!result.startsWith(difficulty));
    return cnt - 1;
  }

  public static void main(String[] args) throws NoSuchAlgorithmException {
    String s = "yzbqklnj";
    System.out.println("Part One = " + partOne(s));
    System.out.println("Part Two = " + partTwo(s));
  }
}

File: src/aoc2015/kot/Day04.kt
package aoc2015.kot

import java.security.MessageDigest

object Day04 {
  fun solve(salt: String, prefix: String): Int = generateSequence(0) { it + 1 }.map { md5Hash(salt + it) }.indexOfFirst { it.startsWith(prefix) }

  private fun md5Hash(input: String): String = MessageDigest.getInstance("MD5")!!.digest(input.toByteArray()).toHex()

  private fun ByteArray.toHex(): String {
    val result = StringBuffer()

    forEach {
      val octet = it.toInt()
      val firstIndex = (octet and 0xF0).ushr(4)
      val secondIndex = octet and 0x0F
      result.append("0123456789ABCDEF"[firstIndex])
      result.append("0123456789ABCDEF"[secondIndex])
    }
    return result.toString()
  }
}

fun main(args: Array<String>) {
  println("Part One = ${Day04.solve("yzbqklnj", "00000")}")
  println("Part Two = ${Day04.solve("yzbqklnj", "000000")}")
}
